## 1. lambda表达式生成类的情况讨论，静态 | 实例，一个方法内有多个lambda表达式C#编译器会怎么处理

```cs
internal class Program
{
	public static void Main(string[] args)
	{
		Func<int, int, int> ac1 = (a, b) => { return a + b; };
		Func<int, int, int, int> ac2 = (a, b, c) => { return a + b + c; };

		Console.WriteLine(ac1(10, 12));
		Console.WriteLine(ac2(10, 12, 14));
	}
}
```

查看Il 发现C#编译器生成了一个私有的嵌套类 (删掉了方法体里的IL代码)

```cs
 .class auto ansi serializable sealed nested private beforefieldinit '<>c'
         extends [System.Runtime]System.Object
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
    .field public static initonly class Lambda.Program/'<>c' '<>9'
    .field public static class [System.Runtime]System.Func`3<int32,int32,int32> '<>9__0_0'
    .field public static class [System.Runtime]System.Func`4<int32,int32,int32,int32> '<>9__0_1'
    .method private hidebysig specialname rtspecialname static 
            void  .cctor() cil managed
    {
    } // end of method '<>c'::.cctor
    .method public hidebysig specialname rtspecialname 
            instance void  .ctor() cil managed
    {
    } // end of method '<>c'::.ctor
    .method assembly hidebysig instance int32 
            '<Main>b__0_0'(int32 a,
                           int32 b) cil managed
    {
    } // end of method '<>c'::'<Main>b__0_0'
    .method assembly hidebysig instance int32 
            '<Main>b__0_1'(int32 a,
                           int32 b,
                           int32 c) cil managed
    {
    } // end of method '<>c'::'<Main>b__0_1'
  } // end of class '<>c'

```

有成员如下：

- 公共静态只读字段 `<>9`，类型为 `Lambda.Program/'<>c'` (自己类的一个静态实例)
- 公共静态字段 `<>9__0_0`，类型为 `System.Func3<int32,int32,int32>`
- 公共静态字段 `<>9__0_1`，类型为 `System.Func4<int32,int32,int32,int32>`
- 实例方法 `<Main>b__0_0` 对应我第一个lambda表达式
- 实例方法 `<Main>b__0_1` 对应我第二个lambda表达式

Main方法IL

```cs
   .method public hidebysig static void  Main(string[] args) cil managed
  {
    .custom instance void System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = ( 01 00 01 00 00 ) 
    // Code size       95 (0x5f)
    .maxstack  4
    .locals init (class [System.Runtime]System.Func`3<int32,int32,int32> V_0)
    IL_0000:  ldsfld     class [System.Runtime]System.Func`3<int32,int32,int32> Lambda.Program/'<>c'::'<>9__0_0'
    IL_0005:  dup
  
    IL_0006:  brtrue.s   IL_001f // 这里有一个流程控制，判断栈上变量是否 true，是否null
    // --------------------- 使得下面这段逻辑 在多次方法调用只会创建一个委托实例
    IL_0008:  pop
    IL_0009:  ldsfld     class Lambda.Program/'<>c' Lambda.Program/'<>c'::'<>9'
    IL_000e:  ldftn      instance int32 Lambda.Program/'<>c'::'<Main>b__0_0'(int32,
                                                                             int32)
    IL_0014:  newobj     instance void class [System.Runtime]System.Func`3<int32,int32,int32>::.ctor(object, native int)
    IL_0019:  dup

    IL_001a:  stsfld     class [System.Runtime]System.Func`3<int32,int32,int32> Lambda.Program/'<>c'::'<>9__0_0'
    // -------------------------
    IL_001f:  stloc.0 // 跳转到这里
    IL_0020:  ldsfld     class [System.Runtime]System.Func`4<int32,int32,int32,int32> Lambda.Program/'<>c'::'<>9__0_1'
    IL_0025:  dup
    IL_0026:  brtrue.s   IL_003f
    IL_0028:  pop
    IL_0029:  ldsfld     class Lambda.Program/'<>c' Lambda.Program/'<>c'::'<>9'
    IL_002e:  ldftn      instance int32 Lambda.Program/'<>c'::'<Main>b__0_1'(int32,int32,int32)
    IL_0034:  newobj     instance void class [System.Runtime]System.Func`4<int32,int32,int32,int32>::.ctor(object, native int)
    IL_0039:  dup
  
    IL_003a:  stsfld     class [System.Runtime]System.Func`4<int32,int32,int32,int32> Lambda.Program/'<>c'::'<>9__0_1'
    IL_003f:  ldloc.0
    IL_0040:  ldc.i4.s   10
    IL_0042:  ldc.i4.s   12
    IL_0044:  callvirt   instance !2 class [System.Runtime]System.Func`3<int32,int32,int32>::Invoke(!0,!1)
    IL_0049:  call       void [System.Console]System.Console::WriteLine(int32)
  
    IL_004e:  ldc.i4.s   10
    IL_0050:  ldc.i4.s   12
    IL_0052:  ldc.i4.s   14
    IL_0054:  callvirt   instance !3 class [System.Runtime]System.Func`4<int32,int32,int32,int32>::Invoke(!0,!1,!2)
    IL_0059:  call       void [System.Console]System.Console::WriteLine(int32)
    IL_005e:  ret
  } // end of method Program::Main
```

### 捕获局部变量

```cs
    internal class Program
    {
        public static void Main(string[] args)
        {
            int d = 10;
            Func<int, int, int> ac1 = (a, b) => { return a + b + d; };
            Func<int, int, int, int> ac2 = (a, b, c) => { return a + b + c + d; };

            Console.WriteLine(ac1(10, 12));
            Console.WriteLine(ac2(10, 12, 14));
        }
    }

```

```cs
.class auto ansi sealed nested private beforefieldinit '<>c__DisplayClass0_0'
         extends [System.Runtime]System.Object
  {
    .custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
    .field public int32 d
    .method public hidebysig specialname rtspecialname 
            instance void  .ctor() cil managed
    {
    } // end of method '<>c__DisplayClass0_0'::.ctor
    .method assembly hidebysig instance int32 
            '<Main>b__0'(int32 a,
                         int32 b) cil managed
    {
    } // end of method '<>c__DisplayClass0_0'::'<Main>b__0'
    .method assembly hidebysig instance int32 
            '<Main>b__1'(int32 a,
                         int32 b,
                         int32 c) cil managed
    {
    } // end of method '<>c__DisplayClass0_0'::'<Main>b__1'
  } // end of class '<>c__DisplayClass0_0'
```

- 公共的32位整数字段 `d` （对应局部变量 d）
- 实例方法 `<Main>b__0`
- 实例方法 `<Main>b__1`

捕获了局部变量，两个静态委托字段和自己类的静态实例就没了

Main方法IL

```cs
.method public hidebysig static void  Main(string[] args) cil managed
  {
    .custom instance void System.Runtime.CompilerServices.NullableContextAttribute::.ctor(uint8) = ( 01 00 01 00 00 ) 
    // Code size       69 (0x45)
    .maxstack  4
    .locals init (class [System.Runtime]System.Func`3<int32,int32,int32> V_0)
    IL_0000:  newobj     instance void Lambda.Program/'<>c__DisplayClass0_0'::.ctor()
    IL_0005:  dup
  
    IL_0006:  ldc.i4.s   10
    IL_0008:  stfld      int32 Lambda.Program/'<>c__DisplayClass0_0'::d
    IL_000d:  dup
    IL_000e:  ldftn      instance int32 Lambda.Program/'<>c__DisplayClass0_0'::'<Main>b__0'(int32, int32)
    IL_0014:  newobj     instance void class [System.Runtime]System.Func`3<int32,int32,int32>::.ctor(object, native int)
  
    IL_0019:  stloc.0
    IL_001a:  ldftn      instance int32 Lambda.Program/'<>c__DisplayClass0_0'::'<Main>b__1'(int32, int32, int32)
    IL_0020:  newobj     instance void class [System.Runtime]System.Func`4<int32,int32,int32,int32>::.ctor(object, native int)
  
    IL_0025:  ldloc.0
    IL_0026:  ldc.i4.s   10
    IL_0028:  ldc.i4.s   12
    IL_002a:  callvirt   instance !2 class [System.Runtime]System.Func`3<int32,int32,int32>::Invoke(!0, !1)
    IL_002f:  call       void [System.Console]System.Console::WriteLine(int32)
  
    IL_0034:  ldc.i4.s   10
    IL_0036:  ldc.i4.s   12
    IL_0038:  ldc.i4.s   14
    IL_003a:  callvirt   instance !3 class [System.Runtime]System.Func`4<int32,int32,int32,int32>::Invoke(!0,!1, !2)
    IL_003f:  call       void [System.Console]System.Console::WriteLine(int32)
    IL_0044:  ret
  } // end of method Program::Main
```

总结：

- 没有捕获上下文的lambda表达式会在嵌套类中生成 静态的实例和静态的委托字段（lambda表达式写的方法）
  - 并且通过一个流程控制语句，使得多次进入方法不会创建相同的委托实例。
- 有捕获上下文（只捕获实例变量，静态变量直接通过类名调用），会将上下文包装到嵌套类中。
- 一个方法内的所有lambda会共享一个嵌套类

## 2. 特性可以传入自定义 Struct 吗

```cs
    struct MyStruct
    {
    }

    [AttributeUsage(AttributeTargets.All, Inherited = false, AllowMultiple = true)]
    sealed class MyAttribute : Attribute
    {
        private MyStruct ms;

        public MyAttribute(MyStruct ms)
        {
            this.ms = ms;
        }
    }

    [My(new MyStruct())] // 这里报错 Attribute constructor parameter 'ms' has type 'AttrMStructCanDo.MyStruct', which is not a valid attribute parameter type
    internal class Program
    {
        public static void Main(string[] args)
        {
        }
    }
```

> 应用特性时必须传递一个编译时常量表达式，它与特性类定义的类型匹配。在特性类定义了一个 `Type`参数、`Type`字段或者 `Type`属性的任何地方，都必须使用 C# `typeof` 操作符(如下例所示)。在特性类定义了一个 `Object` 参数、`Object`字段或者 `Object` 属性的任何地方，都可传递一个 `Int32`、`String` 或其他任何常量表达式(包括 `null`)。如果常量表达式代表值类型，那么在运行时构造特性的实例时会对值类型进行装箱。

 定义默认参数, 编译时会报错
 Error CS0181 : 特性构造函数参数“ms”具有类型“MyStruct”，这不是有效特性参数类型

```cs

    struct MyStruct
    {
    }

    [AttributeUsage(AttributeTargets.All, Inherited = false, AllowMultiple = true)]
    sealed class MyAttribute : Attribute
    {
        private MyStruct ms;

        public MyAttribute(MyStruct ms = default)
        {
            this.ms = ms;
        }
    }

    [My()]
    internal class Program
    {
        public static void Main(string[] args)
        {
        }
    }
```

> 引用MSDN
> 自定义特性定义包括对参数指定无效数据类型的构造函数。 特性只能以某些数据类型作为参数，因为只有这些类型可序列化到程序集的元数据。

特性通常在运行时通过反射来读取，而这些值需要以一种可序列化的形式存储在程序集元数据中：

- 基础类型、字符串、枚举等是可序列化的，可以直接存储在元数据中。
- 自定义值类型可能包含复杂的内部状态或行为，甚至可能包含引用类型字段，这些字段在元数据中无法直接存储。

## 3. 多次调用 GetCustomAttribute 效率一样吗 （有优化吗）

不会有优化，书上没有相关描述

写个测试

```cs
   public class Program
    {
        public static void Main(string[] args)
        {
            BenchmarkRunner.Run<Program>();
        }

        [Benchmark]
        public void Test3()
        {
            var type = typeof(Program);
            type.GetCustomAttributes(true);
            type.GetCustomAttributes(true);
            type.GetCustomAttributes(true);
            type.GetCustomAttributes(true);
        }

        [Benchmark(Baseline = true)]
        public void Test2()
        {
            var type = typeof(Program);
            type.GetCustomAttributes(true);
            type.GetCustomAttributes(true);
        }
    }
```

| Method |     Mean |     Error |    StdDev | Ratio |
| ------ | -------: | --------: | --------: | ----: |
| Test3  | 3.303 us | 0.0141 us | 0.0132 us |  1.98 |
| Test2  | 1.668 us | 0.0098 us | 0.0082 us |  1.00 |

## 4. Attribute.Equal 有实现吗

有默认实现，逻辑是使用反射获取每个字段，然后逐一比较。

```cs

    [SecuritySafeCritical]
    [__DynamicallyInvokable]
    public override bool Equals(object obj)
    {
      if (obj == null)
        return false;
      RuntimeType type = (RuntimeType) this.GetType();
      if ((RuntimeType) obj.GetType() != type)
        return false;
      object obj1 = (object) this;
      FieldInfo[] fields = type.GetFields(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
      for (int index = 0; index < fields.Length; ++index)
      {
        if (!Attribute.AreFieldValuesEqual(((RtFieldInfo) fields[index]).UnsafeGetValue(obj1), ((RtFieldInfo) fields[index]).UnsafeGetValue(obj)))
          return false;
      }
      return true;
    }

    private static bool AreFieldValuesEqual(object thisValue, object thatValue)
    {
      if (thisValue == null && thatValue == null)
        return true;
      if (thisValue == null || thatValue == null)
        return false;
      if (thisValue.GetType().IsArray)
      {
        if (!thisValue.GetType().Equals(thatValue.GetType()))
          return false;
        Array array1 = thisValue as Array;
        Array array2 = thatValue as Array;
        if (array1.Length != array2.Length)
          return false;
        for (int index = 0; index < array1.Length; ++index)
        {
          if (!Attribute.AreFieldValuesEqual(array1.GetValue(index), array2.GetValue(index)))
            return false;
        }
      }
      else if (!thisValue.Equals(thatValue))
        return false;
      return true;
    }
```

## 5. Nullable<> 有什么泛型约束，上面加了什么特性，它的大小是怎么样的？

```cs
  [NonVersionable] // 实现不轻易随版本变化修改
  [__DynamicallyInvokable] // 动态反射调用
  [Serializable] // 可序列化
  [method: NonVersionable] 
  [method: __DynamicallyInvokable] 
  public struct Nullable<T>(T value) where T : struct
  {
	}
```

Nullable<\T>  里面有
bool hasValue
T vlaue

具体的大小，根据T的大小而定（由于字节对齐）
bool 占一个字节，后面要补齐 T的大小
所以 Nullable大小是 2 * sizeof(T);

## 6. int? a = 1; a.GetType();是如何调用的，为什么会返回 int 的type实例

逻辑上来说是：a.GetType() 这里会发生装箱，吧 a 里面的int 拿出来装箱，然后调用装箱后的对象的GetType，因此会返回int type实例

## 7. int? a = null; a.GetType(); 是在哪里抛出的异常？

```cs
int? a = null;
a.GetType();
```

代码实际情况是下面的：

```cs
int? a = 1;
object o = a;
var type = o.GetType();
Console.WriteLine(type);
```

中间部分IL如下

```cs
    // [10 13 - 10 26]
    IL_0009: ldloc.0      // a
    IL_000a: box          valuetype [mscorlib]System.Nullable`1<int32>
    IL_000f: stloc.1      // o

    // [11 13 - 11 36]
    IL_0010: ldloc.1      // o
    IL_0011: callvirt     instance class [mscorlib]System.Type [mscorlib]System.Object::GetType()
    IL_0016: stloc.2      // 'type'
```

`callvirt` 中做的nullreferenceExcption抛出

> `callvirt`
> 该 IL 指令可调用实例方法和虚方法，不能调用静态方法。
> 用 `callvirt` 指令调用实例方法或虚方法，必须指定引用了对象的变量。
> 用 `callvirt` 指令调用非虚实例方法，变量的类型指明了方法的定义类型。
>
> 用 `callvirt` 指令调用虚实例方法，CLR 调查发出调用的对象的实际类型，然后以多态方式调用方法。
> 为了确定类型，发出调用的变量绝不能是 `null`。换言之，编译这个调用时，JIT 编译器会生成代码来验证变量的值是不是 `null`。如果是， `callvirt` 指令的执行速度比 `call` 指令稍慢。注意，即使 `callvirt` 指令调用的是非虚实例方法，也要执行这种 `null`检查。

## 8. Finally 语句为什么会执行，它的机制是怎么样的？

> 定义参考:
> finally 代码块用于清理 try 代码块中分配的任何资源，以及运行即使发生异常也必须执行的任何代码。无论 try 代码块如何退出，控制权始终传递给 finally 代码块。

```cs
    internal class Program
    {
        public static void Main(string[] args)
        {
            try
            {
                Console.WriteLine("In try block");
                throw new Exception();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
            finally
            {
                Console.WriteLine("In finally block");
            }

            Console.WriteLine("After finally block");
        }
    }

```

```cs
 .method public hidebysig static void
    Main(
      string[] args
    ) cil managed
  {
    .entrypoint
    .maxstack 1
    .locals init (
      [0] class [mscorlib]System.Exception e
    )

    // [8 9 - 8 10]
    IL_0000: nop
    .try
    {
      .try
      {
        // [10 13 - 10 14]
        IL_0001: nop

        // [11 17 - 11 51]
        IL_0002: ldstr        "In try block"
        IL_0007: call         void [mscorlib]System.Console::WriteLine(string)
        IL_000c: nop

        // [12 17 - 12 39]
        IL_000d: newobj       instance void [mscorlib]System.Exception::.ctor()
        IL_0012: throw
      } // end of .try
      catch [mscorlib]System.Exception
      {

        // [14 13 - 14 32]
        IL_0013: stloc.0      // e

        // [15 13 - 15 14]
        IL_0014: nop

        // [16 17 - 16 38]
        IL_0015: ldloc.0      // e
        IL_0016: call         void [mscorlib]System.Console::WriteLine(object)
        IL_001b: nop

        // [17 13 - 17 14]
        IL_001c: nop
        IL_001d: leave.s      IL_001f
      } // end of catch

      IL_001f: leave.s      IL_002f
    } // end of .try
    finally
    {

      // [19 13 - 19 14]
      IL_0021: nop

      // [20 17 - 20 55]
      IL_0022: ldstr        "In finally block"
      IL_0027: call         void [mscorlib]System.Console::WriteLine(string)
      IL_002c: nop

      // [21 13 - 21 14]
      IL_002d: nop
      IL_002e: endfinally
    } // end of finally

    // [23 13 - 23 54]
    IL_002f: ldstr        "After finally block"
    IL_0034: call         void [mscorlib]System.Console::WriteLine(string)
    IL_0039: nop

    // [24 9 - 24 10]
    IL_003a: ret

  } // end of method Program::Main

```

主要是从流程控制的角度实现的，try中逻辑执行完没有抛出异常在代码流程中就能跳转到finally块中，如果抛出了异常，被本层的catch块捕获了，catch块执行完成后跳转到finally，如何没被捕获而被外层捕获了，则执行CLR会执行内层的finally语句块，假如没被托管代码捕获，这CLR执行流程中应该也有执行finally语句块的逻辑。

如果不遵守流程控制的逻辑，或者无法执行则可能不会执行finally，如下讨论的。

### 例外的情况 [try catch - C# 的 &#34;finally&#34; 块总是执行吗？ - Stack Overflow --- try catch - Does the C# &#34;finally&#34; block ALWAYS execute? - Stack Overflow](https://stackoverflow.com/questions/3216046/does-the-c-sharp-finally-block-always-execute)

- 任何直接终止进程而不是让其自行温和关闭的操作都将阻止 `finally` 的运行
  1. Environment.FailFast
  2. Power Failure 电源故障
- 出现 StackOverflowException 时不会执行，因为栈中没有空间执行任何更多代码。
- 出现极其罕见的 ExecutingEngineException 时，它也不会被调用。

## 9. throw e 和 throw 的区别

```cs
try {

} catch (Exection e) {
	// throw e;
	throw;
}
```

throw 对应的IL 是
IL_0019: rethrow

throw e 对应的IL 是：
IL_0019: ldloc.0      // e
IL_001a: throw

### 做个测试

```cs
   internal class Program
    {
        public static void Main(string[] args)
        {
            try
            {
                Test();
            }
            catch (Exception e)
            {
                Console.WriteLine(e); // 调试这里的异常
            }

            Console.WriteLine("After finally block");
        }

        static void Test()
        {
            try
            {
                TestInner();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }
            finally
            {
                Console.WriteLine("In finally block");
            }
        }

        static void TestInner()
        {
            throw new Exception();
        }
```

```
System.Exception: 引发类型为“System.Exception”的异常。
   在 WhyFinally.Program.TestInner() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 41
   在 WhyFinally.Program.Test() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 26
   
In finally block

System.Exception: 引发类型为“System.Exception”的异常。
   在 WhyFinally.Program.TestInner() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 41
   在 WhyFinally.Program.Test() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 31
   在 WhyFinally.Program.Main(String[] args) 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 11

```

Test 方法里面catch 块改成 throw e;

```cs
System.Exception: 引发类型为“System.Exception”的异常。
   在 WhyFinally.Program.TestInner() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 40
   在 WhyFinally.Program.Test() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 25
In finally block

System.Exception: 引发类型为“System.Exception”的异常。
   在 WhyFinally.Program.Test() 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 30
   在 WhyFinally.Program.Main(String[] args) 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 11
After finally block
```

可见 throw e; 会被视为重新抛出异常，CLR会重置调用栈的起点。

## 10. 不标记可序列化，抛出异常会怎么样？

```cs
    public class MyException : Exception
    {
        public MyException()
        {
        }

        public MyException(string message) : base(message)
        {
        }

        public MyException(string message, Exception inner) : base(message, inner)
        {
        }
    }

    internal class Program
    {
        public static void Main(string[] args)
        {
            throw new MyException();
        }
    }
```

不标记也能正常抛出

```
未经处理的异常:  WhyFinally.MyException: 引发类型为“WhyFinally.MyException”的异常。
   在 WhyFinally.Program.Main(String[] args) 位置 D:\Study\CLR_via_CSharp\resources\SourceCodes\CLR_via_CSharp\WhyFinally\Program.cs:行号 24
```

### 测试嵌套 AppDemain

```cs
    public class MyException : Exception
    {
        private string s;

        public MyException(string s)
        {
            this.s = s;
        }

        public MyException(string s, string message) : base(message)
        {
            this.s = s;
        }

        public MyException(string s, string message, Exception inner) : base(message, inner)
        {
            this.s = s;
        }
    }

    internal class Program
    {
        public static void Main(string[] args)
        {
            TestControl.Go();
        }
    }

    class TestControl
    {
        public static void Go()
        {
            // 创建新的 AppDomain
            AppDomain newDomain = AppDomain.CreateDomain("NewDomain");

            // 在新 AppDomain 中执行代码
            try
            {
                newDomain.DoCallBack(() =>
                {
                    try
                    {
                        // 在新 AppDomain 中抛出 MyException 异常
                        throw new MyException("MyException", "This is an un-serializable exception");
                    }
                    catch (Exception e)
                    {
                        // 捕捉并打印异常信息
                        Console.WriteLine($"Caught exception in new AppDomain: {e.GetType().Name} - {e.Message}");
                        throw;
                    }
                });
            }
            catch (Exception e)
            {
                // 捕捉传递到原始 AppDomain 中的异常
                Console.WriteLine($"Caught exception in original AppDomain: {e.GetType().Name} - {e.Message}");
            }
            finally
            {
                // 卸载新的 AppDomain
                AppDomain.Unload(newDomain);
            }
        }
    }

```

输出如下：

```cs
Caught exception in new AppDomain: MyException - This is an un-serializable exception

Caught exception in original AppDomain: SerializationException - 程序集“WhyFinally, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null”中的类型“WhyFinally.MyException”未标记为可序列化。
```

运行时报错 SerializationException

修改一下异常类

```cs
  [Serializable]
    public class MyException : Exception
    {
        private string s;

        public MyException(string s)
        {
            this.s = s;
        }

        public MyException(string s, string message) : base(message)
        {
            this.s = s;
        }

        public MyException(string s, string message, Exception inner) : base(message, inner)
        {
            this.s = s;
        }

        protected MyException(SerializationInfo info, StreamingContext context) : base(info, context)
        {
            s = info.GetString("s");
        }

        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);
            info.AddValue("s", s);
        }
    }
```

输出如下：

```cs
Caught exception in new AppDomain: MyException - This is an un-serializable exception
Caught exception in original AppDomain: MyException - This is an un-serializable exception
```

外层可以正常捕获。

## 11 malloc 是否线程安全，C# 的GC的内存分配呢？（NextObjPtr指针）


### malloc

[c - Is malloc thread-safe? - Stack Overflow](https://stackoverflow.com/questions/855763/is-malloc-thread-safe)

From the [GNU C Library Reference Manual](http://www.gnu.org/software/libc/manual/html_mono/libc.html#Basic-Allocation) of `glibc-2.20 [released 2014-09-07]`:

> `void * malloc (size_t size)`
> 
> Preliminary: MT-Safe | ...
> 
> ... 1.2.2.1 POSIX Safety Concepts:
> 
> ... MT-Safe or Thread-Safe functions are safe to call in the presence of other threads. MT, in MT-Safe, stands for Multi Thread.
> 
> Being MT-Safe does not imply a function is atomic, nor that it uses any of the memory synchronization mechanisms POSIX exposes to users. It is even possible that calling MT-Safe functions in sequence does not yield an MT-Safe combination. For example, having a thread call two MT-Safe functions one right after the other does not guarantee behavior equivalent to atomic execution of a combination of both functions, since concurrent calls in other threads may interfere in a destructive way.
> 
> Whole-program optimizations that could inline functions across library interfaces may expose unsafe reordering, and so performing inlining across the GNU C Library interface is not recommended. The documented MT-Safety status is not guaranteed underwhole-program optimization. However, functions defined in user-visible headers are designed to be safe for inlining.

目前 `malloc()` 是线程安全的。

GNU GLIBC-2.23 源码

```c

void *
malloc (size_t len)
{
  struct header *result = NULL;

  /* Determine real implementation if not already happened.  */
  if (__glibc_unlikely (initialized <= 0))
    {
      if (initialized == -1)
        return NULL;

      me ();
    }

  /* If this is not the correct program just use the normal function.  */
  if (not_me)
    return (*mallocp)(len);

  /* Keep track of number of calls.  */
  catomic_increment (&calls[idx_malloc]);
  /* Keep track of total memory consumption for `malloc'.  */
  catomic_add (&total[idx_malloc], len);
  /* Keep track of total memory requirement.  */
  catomic_add (&grand_total, len);
  /* Remember the size of the request.  */
  if (len < 65536)
    catomic_increment (&histogram[len / 16]);
  else
    catomic_increment (&large);
  /* Total number of calls of any of the functions.  */
  catomic_increment (&calls_total);

  /* Do the real work.  */
  result = (struct header *) (*mallocp)(len + sizeof (struct header));
  if (result == NULL)
    {
      catomic_increment (&failed[idx_malloc]);
      return NULL;
    }

  /* Update the allocation data and write out the records if necessary.  */
  update_data (result, len, 0);

  /* Return the pointer to the user buffer.  */
  return (void *) (result + 1);
}
```

里面一堆宏，暂时看不懂😢
### C# 内存分配

> .net 6 源码 gchelpers.cpp
 
```cpp
inline Object* Alloc(size_t size, GC_ALLOC_FLAGS flags)
{
    CONTRACTL {
        THROWS;
        GC_TRIGGERS;
        MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
    } CONTRACTL_END;

    _ASSERTE(!NingenEnabled() && "You cannot allocate managed objects inside the ngen compilation process.");

#ifdef _DEBUG
    if (g_pConfig->ShouldInjectFault(INJECTFAULT_GCHEAP))
    {
        char *a = new char;
        delete a;
    }
#endif

    if (flags & GC_ALLOC_CONTAINS_REF)
        flags &= ~GC_ALLOC_ZEROING_OPTIONAL;

    Object *retVal = NULL;
    CheckObjectSize(size);

    if (GCHeapUtilities::UseThreadAllocationContexts())
    {
        gc_alloc_context *threadContext = GetThreadAllocContext();
        GCStress<gc_on_alloc>::MaybeTrigger(threadContext);
        retVal = GCHeapUtilities::GetGCHeap()->Alloc(threadContext, size, flags);
    }
    else
    {
        GlobalAllocLockHolder holder(&g_global_alloc_lock);
        gc_alloc_context *globalContext = &g_global_alloc_context;
        GCStress<gc_on_alloc>::MaybeTrigger(globalContext);
        retVal = GCHeapUtilities::GetGCHeap()->Alloc(globalContext, size, flags);
    }


    if (!retVal)
    {
        ThrowOutOfMemory();
    }

    return retVal;
}
```

可以看到有中间一个if 语句 选择 使用 线程分配上下文（ThreadAllocationContexts）或者 globalContext。
（**线程分配上下文**指托管堆上预分配的一个空白空间，用于线程的内存分配，用完了再分配，减少申请全局锁的开销。）

判断逻辑如下：
```cpp
    static bool UseThreadAllocationContexts()
    {
        // When running on a single-proc Intel system, it's more efficient to use a single global
        // allocation context for SOH allocations than to use one for every thread.
#if (defined(TARGET_X86) || defined(TARGET_AMD64)) && !defined(TARGET_UNIX)
        return IsServerHeap() || ::g_SystemInfo.dwNumberOfProcessors != 1 || CPUGroupInfo::CanEnableGCCPUGroups();
#else
        return true;
#endif

    }
```

>引用自Github Coplots
>1. **单处理器 Intel 系统**：
    - 在单处理器的 Intel 系统上（x86 或 AMD64 架构），如果不是在 Unix 系统上运行，函数会选择使用全局分配上下文而不是每个线程一个分配上下文。这是因为在单处理器系统上，使用全局分配上下文可以减少上下文切换和锁争用，从而提高效率。
>2. **多处理器或服务器堆**：
    - 如果是服务器堆（`IsServerHeap()` 返回 `true`），或者系统的处理器数量不为 1，或者可以启用 CPU 组（`CPUGroupInfo::CanEnableGCCPUGroups()` 返回 `true`），则函数返回 `true`，表示使用线程分配上下文。
    - 这意味着在多处理器环境中，或者在服务器堆配置下，使用线程分配上下文可以更好地利用多核处理器的并行性，提高内存分配的效率。
>3. **非 Intel 或 Unix 系统**：
      - 在非 Intel 架构或 Unix 系统上，函数默认返回 `true`，表示总是使用线程分配上下文。这是因为在这些环境中，线程分配上下文通常能提供更好的性能。


在使用全局分配上下文进行内存分配，获取全局分配锁。防止多个线程同时修改全局分配上下文，确保线程安全。
```cpp
GlobalAllocLockHolder holder(&g_global_alloc_lock);

^
|

typedef GlobalAllocLock::Holder GlobalAllocLockHolder;

^
|

    typedef Holder<GlobalAllocLock *, GlobalAllocLock::AcquireLock, GlobalAllocLock::ReleaseLock> Holder;

^
|

    // Static helper to acquire a lock, for use with the Holder template.
    static void AcquireLock(GlobalAllocLock *lock)
    {
        WRAPPER_NO_CONTRACT;
        lock->Acquire();
    }

    // Static helper to release a lock, for use with the Holder template
    static void ReleaseLock(GlobalAllocLock *lock)
    {
        WRAPPER_NO_CONTRACT;
        lock->Release();
    }

```

lock 是一个 GlobalAllocLock类

```cpp
class GlobalAllocLock {
    friend struct AsmOffsets;
private:
    // The lock variable. This field must always be first.
    LONG m_lock;

public:
    // Creates a new GlobalAllocLock in the unlocked state.
    GlobalAllocLock() : m_lock(-1) {}

    // Copy and copy-assignment operators should never be invoked
    // for this type
    GlobalAllocLock(const GlobalAllocLock&) = delete;
    GlobalAllocLock& operator=(const GlobalAllocLock&) = delete;

    // Acquires the lock, spinning if necessary to do so. When this method
    // returns, m_lock will be zero and the lock will be acquired.
    void Acquire()
    {
        CONTRACTL {
            NOTHROW;
            GC_TRIGGERS; // switch to preemptive mode
            MODE_COOPERATIVE;
        } CONTRACTL_END;

        DWORD spinCount = 0;
        while(FastInterlockExchange(&m_lock, 0) != -1)
        {
            GCX_PREEMP();
            __SwitchToThread(0, spinCount++);
        }

        assert(m_lock == 0);
    }

    // Releases the lock.
    void Release()
    {
        LIMITED_METHOD_CONTRACT;

        // the lock may not be exactly 0. This is because the
        // assembly alloc routines increment the lock variable and
        // jump if not zero to the slow alloc path, which eventually
        // will try to acquire the lock again. At that point, it will
        // spin in Acquire (since m_lock is some number that's not zero).
        // When the thread that /does/ hold the lock releases it, the spinning
        // thread will continue.
        MemoryBarrier();
        assert(m_lock >= 0);
        m_lock = -1;
    }

    // Static helper to acquire a lock, for use with the Holder template.
    static void AcquireLock(GlobalAllocLock *lock)
    {
        WRAPPER_NO_CONTRACT;
        lock->Acquire();
    }

    // Static helper to release a lock, for use with the Holder template
    static void ReleaseLock(GlobalAllocLock *lock)
    {
        WRAPPER_NO_CONTRACT;
        lock->Release();
    }

    typedef Holder<GlobalAllocLock *, GlobalAllocLock::AcquireLock, GlobalAllocLock::ReleaseLock> Holder;
};

```
貌似是个自旋锁。

[CoreCLR 分配小对象内存-CSDN博客](https://blog.csdn.net/kidwr/article/details/142578644)
## 12 根一定是引用类型吗，栈上的值类型有可能作为根吗

值类型中如果有引用类型字段，会被当作根。
jit编译器有执行环境的上下文，可以决定哪些需要作为根，在哪里可以进行GC

## 13 哪些东西会作为活动根

- 静态变量
- 栈上的引用
- GC句柄表里的 normal pined 对象
- freachable队列
- 字符串留用池

## 14 GC分代 中 2代、1代对象修改字段引用，如何知道修改了

引用书上内容：

> 当 JIT 编译器生成本机(native)代码来修改对象中的一个引用字段时，本机代码会生成对一个 write barrier 方法的调用(译注：write barrier 方法是在有数据向对象写入时执行一些内存管理代码的机制)。
>
> 这个 write barrier 方法检查字段被修改的那个对象是否在第 1 代或第 2 代中，
> 如果在，write barrier 代码就在一个所谓的 code table 中设置一个 bit。card table 为堆中的每 128 字节的数据都准备好了一个 bit。GC 下一次启动时会扫描 card table，了解第 1 代和第 2 代中的哪些对象的字段自上次 GC 以来已被修改。
>
> 任何被修改的对象引用了第 0 代中的一个对象，被引用的第 0 代对象就会在垃圾回收过程中“存活”。
>
> GC 之后，card table 中的所有 bit 都被重置为 0。
>
> 向对象的引用字段中写入时，write barrier 代码会造成少量性能损失(对应地，向局部变量或静态字段写入便不会有这个损失)。另外，如果对象在第 1 代或第 2 代中，性能会损失得稍微多一些。

逻辑上讲，一轮GC过后，1代及以上对象引用的引用字段也必然是1代及以上的。我们就不关心则他们里面字段原来引用的对象。

之后新修改的引用就按照上面按块标记的逻辑进行处理。

## 15. 1代 0代内存上是连续的吗？ 1代 0代 是怎么样标记的，谁标记的，在哪？，0代和1代的内存边界是什么


```cpp
// group the frequently used ones together (need intrumentation on accessors)
class generation
{
public:
    // Don't move these first two fields without adjusting the references
    // from the __asm in jitinterface.cpp.
    alloc_context   allocation_context;
    PTR_heap_segment start_segment;
#ifndef USE_REGIONS
    uint8_t*        allocation_start;
#endif //!USE_REGIONS
    heap_segment*   allocation_segment;
    uint8_t*        allocation_context_start_region;
#ifdef USE_REGIONS
    heap_segment*   tail_region;
    heap_segment*   plan_start_segment;
    // only max_generation could have ro regions; for other generations 
    // this will be 0.
    heap_segment*   tail_ro_region;
#endif //USE_REGIONS
    allocator       free_list_allocator;
    size_t          free_list_allocated;
    size_t          end_seg_allocated;
    BOOL            allocate_end_seg_p;
    size_t          condemned_allocated;
    size_t          sweep_allocated;
    size_t          free_list_space;
    size_t          free_obj_space;
    size_t          allocation_size;
#ifndef USE_REGIONS
    uint8_t*        plan_allocation_start;
    size_t          plan_allocation_start_size;
#endif //!USE_REGIONS

    // this is the pinned plugs that got allocated into this gen.
    size_t          pinned_allocated;
    size_t          pinned_allocation_compact_size;
    size_t          pinned_allocation_sweep_size;
    int             gen_num;

#ifdef DOUBLY_LINKED_FL
    BOOL            set_bgc_mark_bit_p;
    uint8_t*        last_free_list_allocated;
#endif //DOUBLY_LINKED_FL

#ifdef FREE_USAGE_STATS
    size_t          gen_free_spaces[NUM_GEN_POWER2];
    // these are non pinned plugs only
    size_t          gen_plugs[NUM_GEN_POWER2];
    size_t          gen_current_pinned_free_spaces[NUM_GEN_POWER2];
    size_t          pinned_free_obj_space;
    // this is what got allocated into the pinned free spaces.
    size_t          allocated_in_pinned_free;
    size_t          allocated_since_last_pin;
#endif //FREE_USAGE_STATS
};
```

generation类 储存各个代的信息, 例如地址范围和使用的段。

跳转allocation_start的值，就可以快速调整对象的代。

如果GC对短暂堆段（存储0，1的堆段）进行清扫，存活下来的对象会变为下一代的对象，代的allocation_start也会随之调整

> 引用官方文档 [Windows 上的大型对象堆 (LOH) - .NET | Microsoft Learn](https://learn.microsoft.com/zh-cn/dotnet/standard/garbage-collection/large-object-heap)
> 垃圾回收器是分代回收器。 它包含三代：第 0 代、第 1 代和第 2 代。 包含 3 代的原因是，在优化良好的应用中，大部分对象都在第 0 代就清除了。 例如，在服务器应用中，与每个请求相关的分配应在请求完成后清除。 仍存在的分配请求将转到第 1 代，并在那里进行清除。 从本质上讲，第 1 代是新对象区域与生存期较长的对象区域之间的缓冲区。
>
> 新分配的对象构成新一代对象，并隐式地成为第 0 代集合。 但是，如果它们是大型对象，它们将延续到大型对象堆 (LOH)，这有时称为第 3 代。 第 3 代是在第 2 代中逻辑收集的物理生成。
>
> 大型对象属于第 2 代，因为只有在第 2 代回收期间才能回收它们。 回收一代时，同时也会回收它前面的所有代。 例如，执行第 1 代 GC 时，将同时回收第 1 代和第 0 代。 执行第 2 代 GC 时，将回收整个堆。 因此，第 2 代 GC 还可称为“完整 GC”。 本文引用第 2 代 GC 而不是完整 GC，但这两个术语是可以互换的。
>
> 代可提供 GC 堆的逻辑视图。 实际上，对象存在于托管堆段中。 托管堆段是 GC 通过调用 [VirtualAlloc 功能](https://learn.microsoft.com/zh-cn/windows/desktop/api/memoryapi/nf-memoryapi-virtualalloc)代表托管代码在操作系统上保留的内存块。 加载 CLR 时，GC 分配两个初始堆段：一个用于小型对象（小型对象堆或 SOH），一个用于大型对象（大型对象堆）。


[CoreCLR源码探索(二) new是什么 - q303248153 - 博客园](https://www.cnblogs.com/zkweb/p/6266117.html)
[CoreCLR源码探索(三) GC内存分配器的内部实现 - q303248153 - 博客园](https://www.cnblogs.com/zkweb/p/6379080.html)